core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java

```diff
@@ -70,13 +81,29 @@ public boolean allocateUnassigned(RoutingAllocation allocation) {
                 continue;
             }
 
-            IndexMetaData indexMetaData = metaData.index(shard.getIndex());
-            Settings indexSettings = Settings.builder().put(settings).put(indexMetaData.getSettings()).build();
+            final boolean snapshotRestore = shard.restoreSource() != null;
+            Set<String> lastActiveAllocationIds = indexMetaData.activeAllocationIds(shard.id());
+
+            final NodesAndVersions nodesAndVersions;
+            final boolean enoughAllocationsFound;
+
+            if (lastActiveAllocationIds.isEmpty() && indexSettings.getIndexVersionCreated().before(Version.V_3_0_0)) {
```

Comment:
do we need the version check here? it's folded into allocatedPostIndexCreate() ? I think it will be simpler to read if we remove this from the if and add an assert on this, explaining why we expect it like that. Something like:

```
            if (lastActiveAllocationIds.isEmpty()) {
                assert indexSettings.getIndexVersionCreated().before(Version.V_3_0_0) : "trying to allocated a primary with an empty allocation id set, but index is new";
```